iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java

```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
+      } else if (detectionStatus.getCurrentPredicate() instanceof OptionPredicate) {
+        matchPredicate((OptionPredicate) detectionStatus.getCurrentPredicate(), detectionStatus);
+      } else if (detectionStatus.getCurrentPredicate() instanceof MultipleUnorderedOptionsPredicate) {
+        matchPredicate((MultipleUnorderedOptionsPredicate) detectionStatus.getCurrentPredicate(), detectionStatus);
       } else {
-        // Stop argument detection in case the argument does not match and the predicate is not optional or should not be matched
+        detectionStatus.setStatus(CommandDetectionStatus.Status.ABORT);
+      }
+
+      if (detectionStatus.is(CommandDetectionStatus.Status.ABORT, CommandDetectionStatus.Status.FOUND_NOTHING)) {
         return Collections.emptyList();
       }
     }
-    return commandArguments;
+    return detectionStatus.commandArguments;
+  }
+
+  private void matchPredicate(SingularPredicate singularPredicate, CommandDetectionStatus detectionStatus) {
+    ArgumentResolution resolution = detectionStatus.argumentStack.pollFirst();
+
+    if (resolution == null) {
+      // nothing to match, will be caught above
+      detectionStatus.setStatus(CommandDetectionStatus.Status.CONTINUE);
+      return;
+    }
+
+    // Test argument resolution with predicate
+    if (singularPredicate.predicate.test(resolution.value())) {
+      // Skip argument and start new command detection
+      if (singularPredicate.is(NO_MATCH)) {
+        detectionStatus.setStatus(CommandDetectionStatus.Status.ABORT);
+        return;
+      }
+      // Re-add predicate to stack to be reevaluated on the next argument
+      if (singularPredicate.is(ZERO_OR_MORE)) {
+        if (detectionStatus.getCurrentPredicate() instanceof SingularPredicate) {
+          // only needed in this case, the other will handle it themselves
+          detectionStatus.predicateStack.addFirst(singularPredicate);
+        }
+      }
+      // Add matched argument to command
+      detectionStatus.commandArguments.add(resolution);
+    } else if (singularPredicate.is(OPTIONAL, ZERO_OR_MORE, NO_MATCH)) {
+      // Re-add argument to be evaluated by the next predicate
+      detectionStatus.argumentStack.addFirst(resolution);
+    } else {
+      // Stop argument detection in case the argument does not match and the predicate is not optional or should not be matched
+      // above method should return empty
+      detectionStatus.setStatus(CommandDetectionStatus.Status.FOUND_NOTHING);
+      return;
+    }
+    detectionStatus.setStatus(CommandDetectionStatus.Status.CONTINUE);
+  }
+
+  private void matchPredicate(OptionPredicate optionPredicate, CommandDetectionStatus detectionStatus) {
+    matchPredicate(optionPredicate.flagPredicate, detectionStatus);
+    if (detectionStatus.is(CommandDetectionStatus.Status.ABORT, CommandDetectionStatus.Status.FOUND_NOTHING) || optionPredicate.withoutValue()) {
```

Comment:
I would replace `optionPredicate.withoutValue()` with `optionPredicate.valuePredicate == null` as this helps analyzers to see that 166 is not null